### ISSUE TITLE
[quantization] Support Gemma 4 E2B assistant model

### ISSUE THREAD

**Issue by mhs4670go** [Samsung/TICO#965]
## Summary

Add post-training quantization (PTQ) and static export support for [`google/gemma-4-E2B-it-assistant`](https://huggingface.co/google/gemma-4-E2B-it-assistant) in TICO.

The assistant model is not a standalone replacement for `google/gemma-4-E2B-it`. It is a lightweight multi-token prediction drafter used during assisted/speculative decoding. It consumes the target model's hidden state and shared KV states, predicts candidate tokens, and lets the target model verify them.

This issue focuses on quantizing and exporting the assistant-side draft computation. End-to-end target verification, candidate acceptance, and target KV commit logic are explicitly out of scope for this first step.

## Background

The Gemma 4 E2B assistant has a different execution contract from a normal causal language model:

- It receives a concatenation of the target token embedding and the target or previous projected hidden state.
- It reuses the target model's `full_attention` and `sliding_attention` shared KV states.
- All assistant decoder layers are shared-KV consumers and do not build their own KV cache.
- It uses a single-position draft step (`B=1`, `Q=1`) repeatedly to generate multiple candidate tokens.
- Its ordered sparse output head performs centroid scoring, `TopK`, dynamic weight gathering, selected-token matmul, and vocabulary-position scattering.

Because of this structure, treating the assistant as a regular `Gemma4ForCausalLM` would produce an incorrect graph and calibration flow.

## Goal

Provide a reusable TICO flow that can:

1. Load and validate a Gemma 4 assistant checkpoint.
2. Quantize the assistant with WrapQ PTQ.
3. Calibrate it using inputs captured from actual Gemma 4 assisted generation.
4. Export a static-shape assistant core graph suitable for NPU execution.
5. Preserve the ordered sparse output-head semantics without forcing unsupported dynamic operations into the NPU graph.
6. Validate floating-point parity, PTQ accuracy, static-shape parity, and exportability.

### NPU: assistant draft-one core

The first implementation should export one fixed-shape graph containing:

```text
assistant_input
    -> pre_projection
    -> 4-layer Gemma 4 assistant backbone
       - shared full-attention KV input
       - shared sliding-attention KV input
       - no assistant-owned KV cache
    -> final norm
    -> post_projection
    -> projected_state

assistant hidden state
    -> centroid projection
    -> centroid_logits
```

Recommended output tensors:

- `projected_state`: next-step state in the target backbone hidden dimension
- `assistant_hidden`: assistant hidden state used by the sparse token head
- `centroid_logits`: scores for centroid selection

### Host/CPU: ordered sparse token head

Unless the backend is confirmed to support the full dynamic sparse-head graph efficiently, keep the following operations outside the NPU graph:

```text
centroid_logits
    -> TopK centroid indices
    -> token_ordering lookup
    -> selected token IDs

assistant_hidden + selected LM-head rows
    -> selected logits
    -> argmax
    -> next draft token ID
```

This avoids exporting:

- dynamic `TopK`-dependent gather,
- dynamic LM-head row selection,
- full-vocabulary scatter,
- unnecessary dense logits over the complete 262K vocabulary.

The host implementation must preserve the same selected token set and top-1 token as the Hugging Face ordered sparse head.

## Static NPU contract

The initial export target is a fixed `draft-one` graph with:

- batch size: `1`
- query length: `1`
- fixed full-attention KV capacity
- fixed sliding-attention KV capacity
- additive attention masks
- explicit tensor-only inputs and outputs
- no Python dictionaries or Hugging Face `ModelOutput` objects in the exported graph

A possible flattened input ABI is:

```text
assistant_input
full_key
full_value
sliding_key
sliding_value
full_attention_mask
sliding_attention_mask
full_cos
full_sin
sliding_cos
sliding_sin
```

All dimensions must be derived from the assistant and target configurations rather than hardcoded.

A static-input canonicalization helper should:

- validate `B=1` and `Q=1`,
- pad or validate shared KV tensors against configured capacities,
- construct masks for valid and padded KV regions,
- preserve the Hugging Face sliding-KV ordering and mask semantics,
- normalize dtype and device,
- produce the flattened export input tuple.

Inputs that exceed the configured capacity must fail with a clear error instead of being silently cropped.

## Quantization plan

Use a conservative baseline first:

- activations: INT16, per tensor
- linear weights: UINT8, per channel
- `pre_projection`: UINT8 weight
- `post_projection`: UINT8 weight
- attention and MLP linear layers: UINT8 weight
- centroid projection: UINT8 weight
- tied embedding / LM-head weight: UINT8
- norm weights: follow the existing Gemma 4 TICO convention

The default profile should be named or documented as `W8A16`.

The configuration should leave room for a later `W4A16` experiment, while allowing sensitive boundaries such as `pre_projection`, `post_projection`, centroid projection, and the sparse LM head to remain W8.

## Calibration

Do not calibrate the assistant only with random tensors.

The assistant activation distribution depends on:

- target token embeddings,
- target final hidden states,
- previous assistant projected states,
- target full-attention shared KV,
- target sliding-attention shared KV,
- prompt length and padding,
- draft depth within one assisted-decoding round.

Calibration should therefore capture real assistant calls from Hugging Face assisted generation:

1. Load the floating-point target model.
2. Load the floating-point assistant model.
3. Register a forward pre-hook, or equivalent instrumentation, on the assistant.
4. Run `target.generate(..., assistant_model=assistant)` on calibration prompts.
5. Capture the assistant call arguments:
   - `inputs_embeds`
   - `position_ids`
   - `attention_mask`
   - `shared_kv_states`
6. Stream each captured call into the prepared TICO assistant model to update observers.

Large KV tensors should not be accumulated unnecessarily in memory.

Suggested calibration options:

```yaml
calibration:
  dataset: wikitext2
  split: train
  n_prompts: 64
  max_prompt_length: 512
  max_new_tokens: 16
  num_assistant_tokens: 6
  max_assistant_calls: 384
  seed: 42
```

## TICO integration

Add a separate model family rather than overloading the existing Gemma 4 adapter with assistant-specific branches.

Expected integration points include:

- a top-level WrapQ wrapper for `Gemma4AssistantForCausalLM`,
- an assistant-specific backbone wrapper,
- ordered sparse-head utilities,
- static input canonicalization,
- an assistant PTQ config builder,
- a recipe adapter,
- Circle export helpers,
- example quantization and export YAML files,
- inspector or wrapper-smoke coverage,
- a static debug runtime.

Existing Gemma 4 decoder-layer, attention, MLP, RMSNorm, and NPU export conventions should be reused where their semantics match. The existing normal causal-model container should not be reused blindly because it owns token embedding, causal-mask generation, optional PLE, and cache behavior that the assistant does not use.

## Weight tying

The assistant ties `lm_head.weight` to `model.embed_tokens.weight`.

The implementation must:

- keep one source of truth for the tied tensor,
- avoid quantizing or serializing two independent copies,
- preserve or explicitly reconstruct the invariant after checkpoint load,
- reuse the same quantized tensor for the host sparse head,
- add a checkpoint round-trip test for the tied-weight behavior.

## Exported artifacts

The flow should generate at least:

```text
quantized_model.pt
gemma4_assistant_core.q.circle
gemma4_assistant_sparse_head.pt
gemma4_assistant_manifest.json
```

The sparse-head artifact should contain enough information to reproduce candidate scoring on the host, including:

- quantized tied weight,
- scale and zero point,
- quantization axis and dtype,
- `token_ordering`,
- number of centroids,
- centroid `TopK`,
- vocabulary size per centroid,
- assistant hidden size,
- artifact schema version.

The manifest should describe:

- source model,
- Transformers version,
- assistant dimensions,
- input and output names, shapes, and dtypes,
- full and sliding KV capacities,
- attention-mask fill value,
- quantization profile,
- sparse-head execution location,
- artifact filenames,
- tied-weight invariant,
- checksum for integer metadata such as `token_ordering`.

## Non-goals

The following work is not part of this issue:

- target `verify-K` Circle graph,
- candidate acceptance and rejection logic,
- target KV commit, rollback, or crop logic,
- complete end-to-end NPU speculative-decoding orchestration,
- dynamic assistant-token scheduling,
- performance tuning beyond the first functional W8A16 baseline.

These should be handled in follow-up issues after assistant PTQ and static draft-one export are working.

## References

- [Gemma 4 E2B instruction-tuned target model](https://huggingface.co/google/gemma-4-E2B-it)
- [Gemma 4 E2B assistant model](https://huggingface.co/google/gemma-4-E2B-it-assistant)
- [Transformers Gemma 4 assistant implementation](https://github.com/huggingface/transformers/blob/v5.9.0/src/transformers/models/gemma4_assistant/modeling_gemma4_assistant.py)
- [Transformers Gemma 4 assistant configuration](https://github.com/huggingface/transformers/blob/v5.9.0/src/transformers/models/gemma4_assistant/configuration_gemma4_assistant.py)
- [Transformers assisted candidate generator](https://github.com/huggingface/transformers/blob/v5.9.0/src/transformers/generation/candidate_generator.py)
